File: Problem1.java
// "static void main" must be defined in a public class.// Find Missing Number in a sorted array
// Input : arr[] = [1, 2, 3, 5, 6, 7, 8]Output : 4//Input : arr[] = [1, 2, 4, 5, 6, 7, 8, 9]Output : 3
//Time Complexity O(log n)
//Space complexity O(1)
import java.util.*;

public class Problem1 { 
public static void main(String[] args) { 
    int[] arr = {1, 2, 3, 5, 6, 7, 8}; 
    System.out.println(Arrays.toString(arr)); 
    Problem1 p = new Problem1();
    System.out.println(p.findMissing(arr)); 
} 
public static int findMissing(int[] arr){ 
    int low = 0; 
    int high = arr.length -1; 
    int mid = 0; 

    while(high - low >= 2){ //means there is a number missing between high and low
        mid = low + (high - low)/2; 
        int lowIdxDiff = arr[low] - low; 
        int highIdxDiff = arr[high] - high; 
        int midIdxDiff = arr[mid] - mid; //no missing element 
        if(midIdxDiff != lowIdxDiff){ 
            high = mid; 
        }else if(midIdxDiff != highIdxDiff){ 
            low = mid; 
        }
            } 
        return (arr[low] + arr[high])/2;
     }
    }

File: Problem1.cpp
//Question: Find the missing number in the sorted array.

// Time Complexity : O(logn)
// Space Complexity : O(1)
// Any problem you faced while coding this : no.


// Your code here along with comments explaining your approach
// Second approach idea is to use binary search. We first find mid and then calculate number of missing elements between two indexes
// and then adjust the low/high accordingly. At the end return mean of the low and high array elements.

#include <vector>
#include <set>
#include <iostream>

// Helper function to find the missing number
int findMissingNumber(std::vector<int>& v)
{
    int low = 0;
    int high = v.size() - 1;
    int mid = 0;
    
    // Atleast 2 elements are needed to check missing number
    while((high - low) >= 2)
    {
        mid = low + (high - low)/2;
        // Calculate how much mid is off with its element.
        int midIdxDiff = v[mid]- mid;
        // Calculate how much low is off with its element.
        int lowIdxDiff = v[low] - low;
        
        // If lowIdxDiff is different from midIdxDiff which means number is missing from left side of array
        if(lowIdxDiff != midIdxDiff)
        {
            high = mid;
        }
        else // If midIdxDiff is different from highIdxDiff, which means number is missing from right side of array.
        {
            low = mid;
        }
    }
    // return the mean of latest low and high to get the missing number.
    return ((v[low]+v[high])/2);
}

int main()
{
    std::vector<int> vec = {1, 2, 3, 4, 5, 6, 8, 9};
    
    std::cout << "Missing number is - " << findMissingNumber(vec) << std::endl;
}

